docs: update spec for timestamping (#290)
notaryproject/notaryproject: specs/signing-and-verification-workflow.md

### Signing Prerequisites

- User has access to the signing certificate and private key or a remote signing service through a notation plugin.

### Signing Steps

1. **Generate signature:** Using notation CLI or any other compliant signing tool, sign the OCI artifact. The signing tool should follow the following guideline.
    1. Verify that the signing certificate is valid and satisfies [certificate requirements](./signature-specification.md#certificate-requirements).
    1. Verify that the signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).
    1. Generate signature.
        1. Generate signature envelope using signature formats specified in [supported signature envelopes](./signature-specification.md#supported-signature-envelopes). Also, as part of this step, the user-defined/supplied custom attributes should be added to the annotations of the signature's descriptor.

>         1. If signing scheme is [`notary.x509`](./signing-scheme.md/#notaryx509) and timestamp countersignature is demanded, extract the [primitive signature](./signature-specification.md#signature-envelope) (the digital signature computed on payload and signed attributes) from the signature envelope generated in the previous step. Compute hash of the signature (the hash algorithm SHOULD be deduced from signing certificate's public key) and send it to a [RFC-3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant `Timestamp Authority (TSA)` for timestamping. Otherwise, continue to the next step. 

Reviewer: There are two plugin commands for generating signatures: `generate-signature` and `generate-envelop`.

For the `generate-signature` command, I think the notation should create the timestamp countersignature for the signature envelope. But for the `generate-envelop` command, should the notation generate the timestamp countersignature, or should the plugin handle it? If the notation generates the timestamp countersignature, that means the notation modifies the plugin-generated envelope. In that case, we may need to mention in the plugin documentation that Notation may change the unsigned field of the plugin-generated envelope. If the timestamp is generated by the plugin, we should provide a consistent protocol for plugin developers to handle the request to create the timestamp.

Reviewer: Yes, but it may need more details to explain my questions.

Reviewer: I see. I think the notation CLI should provide a flag for setting up the TSA server in the sign command. However, based on your explanation, this flag doesn't work for the plugin with the `generate-envelope` command.

Reviewer: You are right. It is too detailed to discuss it here.

>         1. If signing scheme is [`notary.x509`](./signing-scheme.md/#notaryx509) and timestamp countersignature is demanded, extract the [primitive signature](./signature-specification.md#signature-envelope) (the digital signature computed on payload and signed attributes) from the signature envelope generated in the previous step. Compute hash of the signature (the hash algorithm SHOULD be deduced from signing certificate's public key) and send it to a [RFC-3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant `Timestamp Authority (TSA)` for timestamping. Otherwise, continue to the next step. 

Reviewer 2: `./signing-scheme.md/#notaryx509` --> `./signing-scheme.md#notaryx509`?

>         1. If signing scheme is [`notary.x509`](./signing-scheme.md/#notaryx509) and timestamp countersignature is demanded, extract the [primitive signature](./signature-specification.md#signature-envelope) (the digital signature computed on payload and signed attributes) from the signature envelope generated in the previous step. Compute hash of the signature (the hash algorithm SHOULD be deduced from signing certificate's public key) and send it to a [RFC-3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant `Timestamp Authority (TSA)` for timestamping. Otherwise, continue to the next step. 

Reviewer 2: > Compute hash of the signature (the hash algorithm SHOULD be deduced from signing certificate's public key)

Should have a link to somewhere

>         1. If signing scheme is [`notary.x509`](./signing-scheme.md/#notaryx509) and timestamp countersignature is demanded, extract the [primitive signature](./signature-specification.md#signature-envelope) (the digital signature computed on payload and signed attributes) from the signature envelope generated in the previous step. Compute hash of the signature (the hash algorithm SHOULD be deduced from signing certificate's public key) and send it to a [RFC-3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant `Timestamp Authority (TSA)` for timestamping. Otherwise, continue to the next step. 

Reviewer 2: `RFC-3161` --> `RFC 3161`

>         1. If signing scheme is [`notary.x509`](./signing-scheme.md/#notaryx509) and timestamp countersignature is demanded, extract the [primitive signature](./signature-specification.md#signature-envelope) (the digital signature computed on payload and signed attributes) from the signature envelope generated in the previous step. Compute hash of the signature (the hash algorithm SHOULD be deduced from signing certificate's public key) and send it to a [RFC-3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant `Timestamp Authority (TSA)` for timestamping. Otherwise, continue to the next step. 

Reviewer 2: > `Timestamp Authority (TSA)`

TSA is not a code term. You can use `*` instead. For example, *Timestamp Authority (TSA)*.

>         1. If signing scheme is [`notary.x509`](./signing-scheme.md#notaryx509) and timestamp countersignature is demanded, extract the [primitive signature](./signature-specification.md#signature-envelope) (the digital signature computed on payload and signed attributes) from the signature envelope generated in the previous step. Compute hash of the signature (the hash algorithm SHOULD be deduced from signing certificate's public key. See [algorithm requirements](./signature-specification.md#signature-algorithm-requirements)) and send it to a [RFC 3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant *Timestamp Authority (TSA)* for timestamping. The `certReq` field in the [timestamping request](https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.1) MUST be set to true. Otherwise, continue to the next step. 

Reviewer 3: ```suggestion
        1. If signing scheme is [`notary.x509`](./signing-scheme.md#notaryx509) and timestamp countersignature is demanded, extract the [primitive signature](./signature-specification.md#signature-envelope) (the digital signature computed on payload and signed attributes) from the signature envelope generated in the previous step. Compute hash of the signature (the hash algorithm MUST be deduced from signing certificate's public key. See [algorithm requirements](./signature-specification.md#signature-algorithm-requirements)) and send it to a [RFC 3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant *Timestamp Authority (TSA)* for timestamping. The `certReq` field in the [timestamping request](https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.1) MUST be set to true. Otherwise, continue to the next step. 
```

It should be SHOULD or MUST?

            1. Verify that the timestamp signing certificate satisfies [certificate requirements](./signature-specification.md#certificate-requirements).
            1. Verify that the timestamp signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).

>             1. On success, embed the timestamp countersignature to the *Timestamp Signature* unsigned attribute of the signature envelope. If any of the above timestamping step failed, implementations SHOULD decide whether or not to fail the signing process.

Reviewer 2: suggestion: 
```diff
- implementations SHOULD decide whether or not to fail the signing process
+ implementations MAY fail the signing process
```

>             1. On success, embed the timestamp countersignature to the *Timestamp Signature* unsigned attribute of the signature envelope. If any of the above timestamping step failed, implementations MAY fail the signing process.

Reviewer 4: Nit: https://dictionary.cambridge.org/us/dictionary/english/embed
```suggestion
            1. On success, embed the timestamp countersignature in the *Timestamp Signature* unsigned attribute of the signature envelope. If any of the above timestamping step failed, implementations MAY fail the signing process.
```

>             1. On success, embed the timestamp countersignature into the *Timestamp Signature* unsigned attribute of the signature envelope. If any of the above timestamping step failed, implementations MAY fail the signing process.

Reviewer 3: ```suggestion
            1. On success, embed the timestamp countersignature into the *Timestamp Signature* unsigned attribute of the signature envelope. If any of the above timestamping step failed, implementations MUST fail the signing process.
```

We dont want signing to succeed if user requested timestamp and timestamp generation failed.

Reviewer 2: The reason why it is a `MAY` instead of `MUST` is to cover a scenario of temporary unavailable TSA so that implementation can choose to update the timestamping signature later on.
Given that we have a workaround of not having a timestamp signature when TSA is not available, I think we can change it to `MUST`.

    1. Verify that the signing certificate is valid and satisfies [certificate requirements](./signature-specification.md#certificate-requirements).
    1. Verify that the signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).
    1. Generate signature.

>         1. Generate signature envelope using signature formats specified in [supported signature envelopes](./signature-specification.md#supported-signature-envelopes). Also, as part of this step, the user-defined/supplied custom attributes should be added to the annotations of the signature's descriptor.

Reviewer 2: "Validate and embed the received timestamp countersignature to the *Timestamp Signature* unsigned attribute of the signature envelope." <- Try not to combine two steps into a single one.

    1. Verify that the signing certificate is valid and satisfies [certificate requirements](./signature-specification.md#certificate-requirements).
    1. Verify that the signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).
    1. Generate signature.
        1. Generate signature envelope using signature formats specified in [supported signature envelopes](./signature-specification.md#supported-signature-envelopes). Also, as part of this step, the user-defined/supplied custom attributes should be added to the annotations of the signature's descriptor.

>         1. If the user wants to use timestamp and is signing under signing scheme [`notary.x509`](./signing-scheme.md/#notaryx509), extract the [primitive signature](./signature-specification.md#signature-envelope) (the digital signature computed on payload and signed attributes) from the signature envelope generated in the previous step. Send it to a [RFC-3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant `Timestamp Authority (TSA)` for timestamping. Otherwise, continue to the next step. 

Reviewer 3: ```suggestion
        1. If signing scheme is [`notary.x509`](./signing-scheme.md/#notaryx509) and timestamp signature is present, extract the [signature](./signature-specification.md#signature-envelope) (the digital signature computed on payload and signed attributes) from the signature envelope generated in the previous step. Calculate hash of the signature and send it to a [RFC-3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant `Timestamp Authority (TSA)` for timestamping. Otherwise, continue to the next step. 
```

Also how will we select the hashing algorithm? AFAIK most of TSAs only support SHA2 and SHA1. https://www.digicert.com/blog/best-practices-timestamping

            1. Verify that the timestamp signing certificate satisfies [certificate requirements](./signature-specification.md#certificate-requirements).

>             1. Verify that the timestamp signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).

Reviewer: Those 2 links don't work.

    1. Verify that the signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).
    1. Generate signature.
        1. Generate signature using signature formats specified in [supported signature envelopes](./signature-specification.md#supported-signature-envelopes). Also, as part of this step, the user-defined/supplied custom attributes should be added to the annotations of the signature's descriptor.

>         1. If the user wants to timestamp the signature, obtain an [RFC-3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant timestamp for the raw signature (the digital signatures computed on payload and signed attributes) generated in the previous step. Otherwise, continue to the next step.

Reviewer 5: RFC-3161 compliant timestamp or time-stamp token as described in the RFC?

        1. If signing scheme is [`notary.x509`](./signing-scheme.md/#notaryx509) and timestamp countersignature is demanded, extract the [primitive signature](./signature-specification.md#signature-envelope) (the digital signature computed on payload and signed attributes) from the signature envelope generated in the previous step. Compute hash of the signature (the hash algorithm SHOULD be deduced from signing certificate's public key) and send it to a [RFC 3161](https://datatracker.ietf.org/doc/html/rfc3161.html) compliant *Timestamp Authority (TSA)* for timestamping. Otherwise, continue to the next step. 
            1. Verify that the timestamp signing certificate satisfies [certificate requirements](./signature-specification.md#certificate-requirements).
            1. Verify that the timestamp signing algorithm satisfies [algorithm requirements](./signature-specification.md#signature-algorithm-requirements).

>             1. On success, embed the timestamp countersignature to the *Timestamp Signature* unsigned attribute of the signature envelope. If any of the above timestamping step failed, implementations MAY fail the signing process.

Reviewer 4: Nit:  https://dictionary.cambridge.org/us/dictionary/english/embed
```suggestion
            1. On success, embed the timestamp countersignature in the *Timestamp Signature* unsigned attribute of the signature envelope. If any of the above timestamping step failed, implementations MAY fail the signing process.
```

>             1. On success, embed the timestamp countersignature into the *Timestamp Signature* unsigned attribute of the signature envelope. If any of the above timestamping step failed, implementations MAY fail the signing process.

Reviewer 3: ```suggestion
            1. On success, embed the timestamp countersignature into the *Timestamp Signature* unsigned attribute of the signature envelope. If any of the above timestamping step failed, implementations MUST fail the signing process.
```

>             1. On success, embed the timestamp countersignature to the *Timestamp Signature* unsigned attribute of the signature envelope. If any of the above timestamping step failed, implementations SHOULD decide whether or not to fail the signing process.

Reviewer 2: Same comments as above
